src/build.h

```diff
@@ -224,6 +225,7 @@ struct Builder {
 
   State* state_;
   const BuildConfig& config_;
+  Jobserver* jobserver_;
```

Comment:
Document what it means for `jobserver_` to be NULL.

Reply:
It's never NULL when running the real `ninja`, only for the `ninja_test` binary

Reply:
Great! Add as a comment:

```suggestion
  Jobserver* jobserver_;  // Never NULL running the real ninja, only for the ninja_test binary
```

Reply:
To make the code easier to follow, I recommend to instead make JobServer a base class that does nothing (never enabled), then add a derived PosixJobServer  class. This will allow you to add a Win32JobServer class later, and remove the need for all the `#ifdef _WIN32` section sprinkled over the code.

Reply:
@digit-google I understand class inheritance can work, but in this case I think I would have to fill the base class with no-op functions to get rid of the preprocessor stuff. Would that not be equally ugly?

Reply:
Yes, you would need no-op methods for the base class, but this leads to code that is easier to read and maintain over time. You can have test-specific implementations if you want as well. Since this class is not performance critical, you can do something like that:

1) Add an `std::unique_ptr<Jobserver> jobserver_` member to this class, and default-initialize it with a base/null instance that does nothing. This avoids modifying tests that don't care about the feature at all.

2) Add a `SetJobserver(std::unique_ptr<Jobserver> jobserver)` method to replace the instance's jobserver with a new value. This can be called after constructing a Builder instance where it is needed only. This passes ownership to avoid lifecycle management issues.

3) The rest of the code here can use `jobserver_->xxxx()` without worrying about the rest.

Reply:
I split the class into a basic pair of base and derived classes, now it is down to one `#ifdef _WIN32`. The rest of these tips revolving around `unique_ptr` is a little advanced for me, so I'm pushing what I have and we can take some more time to figure out that bit and how useful it would be.